cpg_utils/hail_batch.py

```diff
@@ -28,6 +35,228 @@
 """
 
 
+_batch: Optional['Batch'] = None
+
+
+def get_batch(
+    name: str | None = None, default_python_image: str | None = None
+) -> 'Batch':
+    """
+    Wrapper around Hail's `Batch` class, which allows to register created jobs
+    This has been migrated (currently duplicated) out of cpg_workflows
+
+    Parameters
+    ----------
+    name : str, optional, name for the batch
+    default_python_image : str, optional, default python image to use
+
+    Returns
+    -------
+    If there are scheduled jobs, return the batch
+    If there are no jobs to create, return None
+    """
+    global _batch  # pylint: disable=global-statement
+    backend: hb.Backend
+    if _batch is None:
+        if get_config()['hail'].get('backend', 'batch') == 'local':
+            logging.info('Initialising Hail Batch with local backend')
+            backend = hb.LocalBackend(
+                tmp_dir=tempfile.mkdtemp('batch-tmp'),
+            )
+        else:
+            logging.info('Initialising Hail Batch with service backend')
+            backend = hb.ServiceBackend(
+                billing_project=get_config()['hail']['billing_project'],
+                remote_tmpdir=dataset_path('batch-tmp', category='tmp'),
+                token=os.environ.get('HAIL_TOKEN'),
+            )
+        _batch = Batch(
+            name=name or get_config()['workflow'].get('name'),
+            backend=backend,
+            pool_label=get_config()['hail'].get('pool_label'),
+            cancel_after_n_failures=get_config()['hail'].get('cancel_after_n_failures'),
+            default_timeout=get_config()['hail'].get('default_timeout'),
+            default_memory=get_config()['hail'].get('default_memory'),
+            default_python_image=default_python_image
+            or get_config()['workflow']['driver_image'],
+        )
+    return _batch
+
+
+class Batch(hb.Batch):
+    """
+    Thin subclass of the Hail `Batch` class. The aim is to be able to register
+    created jobs, in order to print statistics before submitting the Batch.
+    """
+
+    def __init__(
+        self,
+        name,
+        backend,
+        *args,
+        pool_label=None,
+        **kwargs,
+    ):
+        super().__init__(name, backend, *args, **kwargs)
+        # Job stats registry:
+        self.job_by_label = {}
+        self.job_by_stage = {}
+        self.job_by_tool = {}
+        self.total_job_num = 0
+        self.pool_label = pool_label
+        if not get_config()['hail'].get('dry_run') and not isinstance(
+            self._backend, hb.LocalBackend
+        ):
+            self._copy_configs_to_remote()
+
+    def _copy_configs_to_remote(self):
+        """If configs are local files, copy them to remote"""
+        remote_dir = to_path(self._backend.remote_tmpdir) / 'config'
+        config_path = remote_dir / (str(uuid.uuid4()) + '.toml')
+        with config_path.open('w') as f:
```

Comment:
Do we need this, or want to encourage this?

Reply:
I wasn't sure about this - I'd be happy to remove if I knew it wouldn't break anything.

My thinking on this was that analysis-runner already handled generating the run config and posting to GCP, this would only be useful for jobs where we want to use `get_config` but don't have the libraries required to read from GPC?